src/decoder/plugins/UsfDecoderPlugin.cxx

```diff
@@ -0,0 +1,318 @@
+}
+
+static int
+usf_info(void * context, const char * name, const char * value)
+{
+    struct UsfLoaderState * state = ( struct UsfLoaderState * ) context;
+    if (!value) return 0;
+
+    if ( strcmp( name, "_enablecompare" ) == 0)
+        state->enableCompare = 1;
+    else if ( strcmp( name, "_enableFIFOfull" ) == 0)
+        state->enableFIFOFull = 1;
+
+    return 0;
+}
+
+struct UsfTags {
+    double length = -1; // Track duration. -1 represents looping infinitely
+    double fade = 0;   // Fade out duration
+    bool setTags = false;  // If set, set the tags with the below variables
+    const TagHandler * tagHandler;
+    void * handler_ctx;
+};
+
+static double
+get_length_from_string(const char * string)
+{
+    size_t len = strlen(string);
+    int32_t total = 0; // Total time in milliseconds
+    int32_t finalMult = 1000; // Multiplier for the final unit. If no delimiter is used, use seconds.
+    int32_t localMult = 1; // Multiplier in the currently parsed unit (millisecond, second, minute).
+    int32_t tmp = 0; // 
+    for (int i = len - 1; i >= 0; i--) {
+        char c = string[i];
+        if (c >= '0' && c <= '9') {
+            tmp += (c - '0') * localMult;
+            localMult *= 10;
+        } else {
+            localMult = 1;
+            int mult = 0;
+            if (c == '.') {
+                mult = 1;
+            } else if (c == ':') {
+                mult = 1000;
+                finalMult = 60000;
+            } else {
+                return -1; // Error in parsing duration. Return -1 (looping).
+            }
+            total += tmp * mult;
+            tmp = 0;
+        }
+    }
+    total += finalMult*tmp;
+
+    return total/1000;
+}
+
+static bool
+usf_set_tag(const char * field, const char * name, const char *value, TagType type, UsfTags * tags)
+{
+    if (strcmp(name, field) == 0) {
+        tag_handler_invoke_tag(*tags->tagHandler, tags->handler_ctx, type, value);
+        return true;
+    }
+    return false;
+}
+
+static int
+usf_tags_target(void * context, const char * name, const char * value)
+{
+    struct UsfTags * tags = ( struct UsfTags * ) context;
+    if (!value) return 0;
+
+    if (tags->setTags) {
+        usf_set_tag("title", name, value, TAG_TITLE, tags) ||
+            usf_set_tag("artist", name, value, TAG_ARTIST, tags) ||
+            usf_set_tag("composer", name, value, TAG_COMPOSER, tags) ||
+            usf_set_tag("game", name, value, TAG_ALBUM, tags) ||
+            usf_set_tag("year", name, value, TAG_DATE, tags) ||
+            usf_set_tag("genre", name, value, TAG_GENRE, tags) ||
+            usf_set_tag("track", name, value, TAG_TRACK, tags);
+    }
+
+    if (strcmp("length", name) == 0) {
+        tags->length = get_length_from_string(value);
+    } else if (strcmp("fade", name) == 0) {
+        tags->fade = get_length_from_string(value);
+    }
+    return 0;
+}
+
+
+inline static double
+track_lengths(double length, double fade)
+{
+    return length + fade;
```

Comment:
What does that formula mean? (If that's not obvious, a code comment is missing.)